dynamic-ssl/src/main/java/org/wildfly/security/dynamic/ssl/DynamicSSLContextSpiImpl.java

```diff
@@ -0,0 +1,119 @@
+
+    private final DynamicSSLContextSPI dynamicSSLContextImpl;
+    private final SSLContext configuredDefaultSSLContext;
+    private volatile SSLSocketFactory sslSocketFactory;
+
+    DynamicSSLContextSpiImpl(SSLContext fallbackSslContext) throws NoSuchAlgorithmException {
+        SSLContext configuredDefaultSSLContextTemp;
+        Iterator<DynamicSSLContextSPI> dynamicSSLContextSPIIterator = ServiceLoader.load(DynamicSSLContextSPI.class).iterator();
+        if (dynamicSSLContextSPIIterator.hasNext()) {
+            dynamicSSLContextImpl = dynamicSSLContextSPIIterator.next();
+            configuredDefaultSSLContextTemp = dynamicSSLContextImpl.getConfiguredDefault() == null ? SSLContext.getDefault() : dynamicSSLContextImpl.getConfiguredDefault();
+        } else {
+            dynamicSSLContextImpl = null;
+            configuredDefaultSSLContextTemp = fallbackSslContext;
+        }
+        this.configuredDefaultSSLContext = configuredDefaultSSLContextTemp;
+    }
+
+    @Override
+    protected void engineInit(KeyManager[] keyManagers, TrustManager[] trustManagers, SecureRandom secureRandom) {
+        // ignore
+    }
+
+    @Override
+    protected SSLSocketFactory engineGetSocketFactory() {
+        if (dynamicSSLContextImpl == null) {
+            return configuredDefaultSSLContext.getSocketFactory();
+        }
+        if (sslSocketFactory == null) {
+            synchronized (this) {
+                if (sslSocketFactory == null) {
+                    sslSocketFactory = new DynamicSslSocketFactory(configuredDefaultSSLContext.getSocketFactory(), dynamicSSLContextImpl);
+                }
+            }
+        }
+        return sslSocketFactory;
+    }
+
+    @Override
+    protected SSLServerSocketFactory engineGetServerSocketFactory() {
+        return this.configuredDefaultSSLContext.getServerSocketFactory();
+    }
+
+    @Override
+    protected SSLEngine engineCreateSSLEngine() {
+        return this.configuredDefaultSSLContext.createSSLEngine();
+    }
+
+    @Override
+    protected SSLEngine engineCreateSSLEngine(String host, int port) throws IllegalStateException {
+        if (dynamicSSLContextImpl == null) {
+            return configuredDefaultSSLContext.createSSLEngine(host, port);
+        }
+        try {
+            SSLContext sslContext = dynamicSSLContextImpl
+                    .getSSLContext(new URI(null, null, host, port, null, null, null));
+            if (sslContext == null) {
+                throw new IllegalStateException("Received SSLContext from DynamicSSLContextProvider was null");
+            }
+            return sslContext.createSSLEngine(host, port);
+        } catch (URISyntaxException | DynamicSSLContextException e) {
+            throw new IllegalStateException(e);
+        }
+    }
+
+    @Override
+    protected SSLSessionContext engineGetServerSessionContext() {
+        throw new UnsupportedOperationException("Dynamic SSLContext does not support sessions");
```

Comment:
yes it is also possible to delegate to the corresponding method from configuredDefaultSSLContext. I though this might be better, as sessions are not supported with context that can change dynamically? If I return `configuredDefaultSSLContext`'s session context while user is not using configuredDefaultSSLContext but some other one, it might lead to unexpected behaviour.